{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data mining in biological engineering\n",
    "Doug White"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence alignment with BioPython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### loading sequence into BioPython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "from Bio import SeqIO\n",
    "\n",
    "#Load the sequence\n",
    "file_path = \"E:/Data Science Biology Tutorial/sequence 1.fasta\" #<- INSERT YOUR PATH HERE\n",
    "num_sequences = 0\n",
    "for seq_record in SeqIO.parse(file_path, \"fasta\"):\n",
    "    sequence = seq_record.seq\n",
    "    num_sequences += 1\n",
    "\n",
    "print(num_sequences)\n",
    "print(len(sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### performing sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Subest the sequence\n",
    "from Bio import pairwise2\n",
    "seq_1 = sequence[0:1000]\n",
    "seq_2 = sequence[-1000:-1]\n",
    "\n",
    "#Perform the alignment\n",
    "alignments = pairwise2.align.globalxx(seq_1, seq_2)\n",
    "print(len(alignments))\n",
    "print(alignments[0])\n",
    "\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Use case\n",
    "- Companion diagnostic:\n",
    "    - Patient selection\n",
    "      * by sequencing\n",
    "        - e.g. a boil\n",
    "        - could tell you if a sample is antibiotic resistance\n",
    "- Find \"probes\" for antibiotic resistance, same approach applies for cancer, infectious disease, flu, viral disease: all of that requires"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Accessing the NCBIâ€™s online API with the Python requests library"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### executing a GET request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "import requests\n",
    "\n",
    "#get the webpage\n",
    "web_page = \"https://www.ncbi.nlm.nih.gov/\"\n",
    "response = requests.get(web_page)\n",
    "print(response)\n",
    "print(response.text)\n",
    "\n",
    "#post to the webpage\n",
    "#get the webpage\n",
    "web_page = \"https://www.ncbi.nlm.nih.gov/\"\n",
    "response = requests.post(web_page, data=[])\n",
    "print(response)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### performing a search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#get the webpage\n",
    "web_page = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?\"\n",
    "db = 'db=nuccore'\n",
    "term = \"term=mrsa\"\n",
    "retmax = \"retmax=100\"\n",
    "response = requests.get(web_page + db + '&' + term + '&' + retmax)\n",
    "print(response)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parsing the html response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#get the webpage\n",
    "web_page = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?\"\n",
    "db = 'db=nuccore'\n",
    "term = \"term=mrsa\"\n",
    "retmax = \"retmax=100\"\n",
    "response = requests.get(web_page + db + '&' + term + '&' + retmax)\n",
    "print(response)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Parse the response using lxml\n",
    "from lxml import html\n",
    "\n",
    "tree = html.fromstring(response.content)\n",
    "ids = tree.xpath('//id/text()')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### retrieving the FASTA files using efetch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Fetech the data using efetch\n",
    "web_page = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "db = 'db=nuccore'\n",
    "id_lookup = \"id=\"\n",
    "for uid in ids:\n",
    "    id_lookup += uid +\",\"\n",
    "#Remove the last ,\n",
    "id_lookup = id_lookup[:-1]\n",
    "\n",
    "ret_type = \"rettype=fasta&retmode=text\"\n",
    "#Now fecth the data\n",
    "response = requests.post(web_page + db + '&' + id_lookup + '&' + ret_type)\n",
    "print(response)\n",
    "print(response.text[0:1000])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Storing our data in a databse in the cloud with MongoDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### connecting to the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "from pymongo import MongoClient\n",
    "\n",
    "#Connect to the database\n",
    "mongo_uri = 'mongodb://test_user:Emory18!@ds023455.mlab.com:23455/heroku_kn66qwrg'\n",
    "client = MongoClient(mongo_uri)\n",
    "database = client[\"heroku_kn66qwrg\"]\n",
    "#Make a new collection\n",
    "your_name = \"Doug\" #<- INSERT YOUR NAME HERE\n",
    "data = {} \n",
    "database[your_name + \"_sequences\"].insert_one(data)\n",
    "\n",
    "print(database[your_name + \"_sequences\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### store the first 100 FASTA files from NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "import requests\n",
    "\n",
    "#get the webpage\n",
    "web_page = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?\"\n",
    "db = 'db=nuccore'\n",
    "term = \"term=mrsa\"\n",
    "retmax = \"retmax=100\"\n",
    "response = requests.get(web_page + db + '&' + term + '&' + retmax)\n",
    "print(response)\n",
    "print(response.text)\n",
    "\n",
    "#Parse the response using lxml\n",
    "from lxml import html\n",
    "\n",
    "tree = html.fromstring(response.content)\n",
    "ids = tree.xpath('//id/text()')\n",
    "\n",
    "#Fetech the data using efecth\n",
    "web_page = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "db = 'db=nuccore'\n",
    "id_lookup = \"id=\"\n",
    "for uid in ids:\n",
    "    id_lookup += uid +\",\"\n",
    "#Remove the last ,\n",
    "id_lookup = id_lookup[:-1]\n",
    "\n",
    "ret_type = \"rettype=fasta&retmode=text\"\n",
    "#Now fecth the data\n",
    "response = requests.post(web_page + db + '&' + id_lookup + '&' + ret_type)\n",
    "print(response)\n",
    "\n",
    "#Create a temporary local file to hold the information\n",
    "f = open('temp.fasta', 'w')\n",
    "f.write(response.text)\n",
    "f.close()\n",
    "\n",
    "#load the FASTA files using biopython\n",
    "from Bio import SeqIO\n",
    "\n",
    "#Now save these to mongoDB\n",
    "for seq_record in SeqIO.parse('temp.fasta', \"fasta\"):\n",
    "    seq_data = dict()\n",
    "    seq_data[\"sequence\"] = str(seq_record.seq)\n",
    "    seq_data[\"record_id\"] = seq_record.id\n",
    "    result = database[your_name + \"_sequences\"].insert_one(seq_data)\n",
    "    print(result.inserted_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### accessing the same files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "from pymongo import MongoClient\n",
    "\n",
    "#Connect to the database\n",
    "mongo_uri = 'mongodb://test_user:Emory18!@ds023455.mlab.com:23455/heroku_kn66qwrg'\n",
    "client = MongoClient(mongo_uri)\n",
    "database = client[\"heroku_kn66qwrg\"]\n",
    "#Make a new collection\n",
    "your_name = \"Doug\" #<- INSERT YOUR NAME HERE\n",
    "data = {} \n",
    "database[your_name + \"_sequences\"].insert_one(data)\n",
    "\n",
    "print(database[your_name + \"_sequences\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## setting up a secure database with the Flask library"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### building a simple Flask application"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "from pymongo import MongoClient\n",
    "from flask import Flask\n",
    "\n",
    "#Define the app\n",
    "app = Flask(__name__)\n",
    "\n",
    "#Create our sequence method\n",
    "@app.route('/sequences')\n",
    "def sequences():\n",
    "    pass\n",
    "\n",
    "#run the app locally\n",
    "if __name__ == \"__main__\":\n",
    "    app.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### connect Flask to MongoDb database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import statements\n",
    "from pymongo import MongoClient\n",
    "from flask import Flask\n",
    "from flask import jsonify\n",
    "\n",
    "#Connect to the database\n",
    "mongo_uri = 'mongodb://test_user:Emory18!@ds023455.mlab.com:23455/heroku_kn66qwrg'\n",
    "client = MongoClient(mongo_uri)\n",
    "database = client[\"heroku_kn66qwrg\"]\n",
    "\n",
    "#Define the app\n",
    "app = Flask(__name__)\n",
    "\n",
    "#Create our sequence method\n",
    "@app.route('/sequences')\n",
    "def sequences():\n",
    "    data = []\n",
    "    #Get all of the records in your collection\n",
    "    your_name = \"Doug\"\n",
    "    results = database[your_name + \"_sequences\"].find({})\n",
    "    #Print out the information\n",
    "    for result in results:\n",
    "        del result[\"_id\"]\n",
    "        data.append(result)\n",
    "    return jsonify(data), 200\n",
    "\n",
    "#run the app locally\n",
    "if __name__ == \"__main__\":\n",
    "    app.run()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
